### ISSUE TITLE
No event fired when image is edited 

### ISSUE THREAD

**Issue by rovansteen** [pqina/filepond#187]
### Summary
When using an image editor to edit the image (like Doka) it doesn't fire an event to let you know that the image has been edited.

### Expected behaviour
I expect a custom event to be called like `onfileedit` or just `onfileadd` for a more immutable approach with the file as argument.

### Additional information
Because I want to handle the upload process myself I manually call `filepond.processFile` which works fine for just adding images, but if you edit them there's no event to track that and upload the edited image.

### Versions
```
"filepond": "^3.5.1",
"filepond-plugin-image-crop": "^2.0.0",
"filepond-plugin-image-edit": "^1.0.1",
"filepond-plugin-image-exif-orientation": "^1.0.3",
"filepond-plugin-image-preview": "^3.1.4",
"filepond-plugin-image-resize": "^2.0.1",
"filepond-plugin-image-transform": "^3.1.0",
```

**Comment by rikschennink** [comment]
Does `onpreparefile` no fit this need?

**Comment by rovansteen** [comment]
@rikschennink `onpreparefile` doesn't seem to be called when I edit an image

**Comment by rikschennink** [comment]
Alright, will look into this after the holidays.

**Comment by rikschennink** [comment]
I've just tested with the set up below but can't reproduce the issue. `onpreparefile` is called twice, once when the image is added and once when the image is edited.

Do you have a public test case?

```js
FilePond.registerPlugin(
    FilePondPluginImageExifOrientation,
    FilePondPluginImageCrop,
    FilePondPluginImageResize,
    FilePondPluginImagePreview,
    FilePondPluginImageTransform,
    FilePondPluginImageEdit
);

var pond = FilePond.create(
    document.querySelector('input'), 
    {
        onpreparefile: (file, output) => {
            console.log(file, output)
        },
        imageEditEditor: Doka.create()
    }
);
```

**Comment by rovansteen** [comment]
@rikschennink not sure how to create a public test case like codepen with Doka. But I've narrowed it down to whether or not `instantUpload` is true or false. If it's true it works like you described but if it's set to false it doesn't trigger the `onpreparefile` event. 

Like I explained in the OP I want to trigger the uploading myself as I need to be in control of the promise, that's why I need to set the `instantUpload` config to false.

**Comment by rikschennink** [comment]
Alright, that makes it easier to reproduce, will test again on Monday.

**Comment by rovansteen** [comment]
Thanks! It does seem to work fine the first time I edit an image. Unfortunately the second time I edit the same image the callback is not triggered and I'm getting the following error:
```

filepond-plugin-image-edit.js:172 Uncaught TypeError: Cannot read property 'crop' of undefined
    at editor.onconfirm (filepond-plugin-image-edit.js:172)
    at editor.onconfirm (filepond-plugin-image-edit.js:178)
    at doka.esm.min.js:4715
    at doka.esm.min.js:4716
```

```
        editor.onconfirm = function (_ref4) {
          var data = _ref4.data;
          var crop = data.crop; // update crop metadata

          item.setMetadata({
            crop: crop
          }); // call

          onouterconfirm(data, createItemAPI(item)); // used in instant edit mode

          if (!handleEditorResponse) return;

          editor.onclose = function () {
            handleEditorResponse(true);
            editor.onclose = null;
          };
        };
```

It seems that `data` here is undefined. This seems to only affect the callback, the editor itself works fine and it crops the image properly.

**Comment by rikschennink** [comment]
@rovansteen looking into it now.

**Comment by rovansteen** [comment]
Thanks, that did get rid of the error. But now it seems that every "even" (2nd, 4th, 6th etc.) edit does not upload the file. In short, this is what I'm doing:

```
Doka.create({
  onconfirm: (output, item) => {
    this.pond.processFile(item);
  }
});
```
It does not trigger a network request and there's no error. It does show the button to manually upload the image.

What might be helpful is that at every edit that fails to upload it does execute a `DELETE` request, while on the attempts where it successfully uploads the file it does not.

**Comment by rikschennink** [comment]
Can you try to call `processFile` with the item `id` instead of the object?

**Comment by rovansteen** [comment]
No luck. Also the ID seems to be the same every time, not sure if that's intended?

**Comment by rovansteen** [comment]
Also if I look at the promise that `processFile()` returns, the _even_ ones are never resolved, neither rejected.

**Comment by rikschennink** [comment]
@rovansteen, okay, back to the drawing board.

**Comment by rikschennink** [comment]
I just successfully tested with this setup (fake `process` method).

`item.id` or `item` indeed doesn't matter, both function correctly with the code below.

```js
FilePond.registerPlugin(
    FilePondPluginImageExifOrientation,
    FilePondPluginImageCrop,
    FilePondPluginImageResize,
    FilePondPluginImagePreview,
    FilePondPluginImageTransform,
    FilePondPluginImageEdit
);

const pond = FilePond.create(
    document.querySelector('input.foo'), 
    {
        instantUpload: false,

        server: {
            // fake process method for testing only
            process: (name, file, metadata, load) => {
                setTimeout(() => {
                    load(Date.now());
                }, 1000);
            }
        },

        imageEditEditor: Doka.create({
            onconfirm: (output, item) => {
                console.log('confirmed crop on', item.id);

                pond.processFile(item).then(() => {
                    console.log('done');
                });
            },
            oncancel: (item) => {
                console.log('cancelled', item.id);
            }
        })
    }
);
```

**Comment by rovansteen** [comment]
Hmm that's weird. This is the code I have now.

```jsx
import React from 'react';
import 'doka/doka.min.css';
import * as Doka from 'doka';
import cookie from 'js-cookie';
import Box from 'components/box';
import styled from 'library/styled';
import * as config from 'library/config';
import { File, FilePond, registerPlugin } from 'react-filepond';

import 'filepond/dist/filepond.min.css';
import FilePondPluginImageEdit from 'filepond-plugin-image-edit';
import FilePondPluginImageCrop from 'filepond-plugin-image-crop';
import FilePondPluginImageResize from 'filepond-plugin-image-resize';
import FilePondPluginImagePreview from 'filepond-plugin-image-preview';
import 'filepond-plugin-image-edit/dist/filepond-plugin-image-edit.css';
import FilePondPluginImageTransform from 'filepond-plugin-image-transform';
import 'filepond-plugin-image-preview/dist/filepond-plugin-image-preview.css';
import FilePondPluginFileValidateSize from 'filepond-plugin-file-validate-size';
import FilePondPluginImageExifOrientation from 'filepond-plugin-image-exif-orientation';

registerPlugin(
  FilePondPluginImageExifOrientation,
  FilePondPluginImagePreview,
  FilePondPluginImageCrop,
  FilePondPluginImageResize,
  FilePondPluginImageTransform,
  FilePondPluginImageEdit,
  FilePondPluginFileValidateSize,
);

const Wrapper = styled(Box)`
  .filepond--root {
    margin-bottom: 0;
  }

  .filepond--file-info {
    display: none;
  }
`;

export type ImageUploadVariant = 'circle';

export interface ImageUploadConfigProps {
  required?: boolean;
  width?: number | string;
  height?: number | string;
  variant?: ImageUploadVariant;
  cropAspectRatio?: string | number | null;
}

export interface ImageUploadProps extends ImageUploadConfigProps {
  value?: any;
  onChange: (value: any) => void;
  onUpload: (upload: Promise<any>) => void;
}

interface ImageUploadState {
  image: any;
}

export default class ImageUpload extends React.Component<
  ImageUploadProps,
  ImageUploadState
> {
  /**
   * Define the default props.
   */
  static defaultProps = {
    cropAspectRatio: null,
  };

  /**
   * Get the derived state from the props.
   */
  static getDerivedStateFromProps(
    props: ImageUploadProps,
    state: ImageUploadState,
  ) {
    return {
      image: state.image || props.value,
    };
  }

  /**
   * Define the initial state.
   */
  state = {
    image: null,
  };

  /**
   * Ref to the pond instance.
   */
  pond: any = null;

  /**
   * Server config.
   */
  getConfig() {
    const token = cookie.get('XSRF-TOKEN');

    if (!token) {
      throw new Error('Could not set up a secure connection with the server.');
    }

    return {
      url: config.api.uploadImage,
      process: {
        withCredentials: true,
        headers: {
          Accept: 'application/json',
          'X-XSRF-TOKEN': token,
        },
      },
    };
  }

  /**
   * Handle when a file is added.
   */
  handleAddFile = (error: any, file: any) => {
    if (error) {
      return console.error(error);
    }

    // Origin '3' means that the file originates
    // from the server so no need to upload.
    if (file.origin === 3) {
      return;
    }

    this.uploadFile(file);
  };

  /**
   * Handle when the image is uploaded.
   *
   * Only when the image is succesfully uploaded to the server
   * we consider the value of the upload field changed and
   * call the onChange prop.
   */
  handleUpload = (error: any, file: any) => {
    if (error) throw error;
    this.props.onChange(file.serverId);
  };

  /**
   * Handle when the image is added.
   */
  handleUpdate = (files: Array<any>) => {
    this.setState({ image: files.length ? files[0].file : null });
  };

  /**
   * Handle when the image is removed.
   */
  handleRemoval = () => {
    this.props.onChange(null);
  };

  /**
   * Handle when the file is editted by the filepond image editor.
   */
  handleEdit = (edits: any, file: any) => {
    this.uploadFile(file);
  };

  /**
   * Upload a filepond file to the server.
   */
  uploadFile(file: any) {
    // Upload the file
    const promise = this.pond.processFile(file.id);

    if (this.props.onUpload) {
      this.props.onUpload(promise);
    }
  }

  /**
   * Get the props based on the variant.
   */
  getVariantProps(variant?: string) {
    switch (variant) {
      case 'circle':
        return {
          stylePanelLayout: 'circle compact',
          styleButtonProcessItemPosition: 'center bottom',
          styleButtonRemoveItemPosition: 'center bottom',
          styleLoadIndicatorPosition: 'center bottom',
          styleProgressIndicatorPosition: 'center bottom',
          styleImageEditButtonEditItemPosition: 'right bottom',
        };
      default:
        return {};
    }
  }

  /**
   * Render the component.
   */
  render() {
    const { image } = this.state;
    const { cropAspectRatio, required, width } = this.props;

    return (
      <Wrapper width={width}>
        <FilePond
          name="payload"
          ref={(ref: any) => (this.pond = ref)}
          required={required}
          instantUpload={false}
          imageEditEditor={Doka.create({
            onconfirm: this.handleEdit,
          })}
          maxFileSize="10MB"
          server={this.getConfig()}
          onaddfile={this.handleAddFile}
          onprocessfile={this.handleUpload}
          onupdatefiles={this.handleUpdate}
          onremovefile={this.handleRemoval}
          labelIdle={`Drag & Drop your picture or <span class="filepond--label-action">Browse</span>`}
          {...this.getVariantProps(this.props.variant)}
        >
          {image && <File key={image} src={image} origin="local" />}
        </FilePond>
      </Wrapper>
    );
  }
}
```

I can't really think of anything in my setup that could conflict with this. Do you maybe have any ideas?

**Comment by rikschennink** [comment]
Thanks that would be super helpful.

**Comment by rovansteen** [comment]
This is the most minimal set up that I could set up that has this issue:

```jsx
import React from 'react';
import 'doka/doka.min.css';
import * as Doka from 'doka';
import { FilePond, registerPlugin } from 'react-filepond';

import 'filepond/dist/filepond.min.css';
import FilePondPluginImageEdit from 'filepond-plugin-image-edit';
import FilePondPluginImageCrop from 'filepond-plugin-image-crop';
import FilePondPluginImageResize from 'filepond-plugin-image-resize';
import FilePondPluginImagePreview from 'filepond-plugin-image-preview';
import 'filepond-plugin-image-edit/dist/filepond-plugin-image-edit.css';
import FilePondPluginImageTransform from 'filepond-plugin-image-transform';
import 'filepond-plugin-image-preview/dist/filepond-plugin-image-preview.css';
import FilePondPluginFileValidateSize from 'filepond-plugin-file-validate-size';
import FilePondPluginImageExifOrientation from 'filepond-plugin-image-exif-orientation';

registerPlugin(
  FilePondPluginImageExifOrientation,
  FilePondPluginImagePreview,
  FilePondPluginImageCrop,
  FilePondPluginImageResize,
  FilePondPluginImageTransform,
  FilePondPluginImageEdit,
  FilePondPluginFileValidateSize,
);

export default class ImageUpload extends React.Component {
  /**
   * Handle when the file is editted by the filepond image editor.
   */
  handleEdit = (edits: any, file: any) => {
    this.uploadFile(file);
  };

  /**
   * Upload a filepond file to the server.
   */
  uploadFile(file: any) {
    this.pond.processFile(file);
  }

  /**
   * Render the component.
   */
  render() {
    return (
      <FilePond
        instantUpload={false}
        server={{
          process: (name: any, file: any, metadata: any, load: any) => {
            setTimeout(() => {
              load(Date.now());
            }, 1000);
          },
        }}
        ref={(ref: any) => (this.pond = ref)}
        imageEditEditor={Doka.create({
          onconfirm: this.handleEdit,
        })}
      />
    );
  }
}
```

I don't see much differences with your setup, could it be something related to the React version of Filepond?

**Comment by rovansteen** [comment]
I've dived in a bit and one thing I noticed is that on this line https://github.com/pqina/filepond/blob/f75e733135955b80aa78ef05ca5161648d1de5ba/dist/filepond.esm.js#L4034 it checks if the item is ready to be processed and if not it returns. On the failed (even) attempts `item.status` is `ItemStatus.PROCESSING_COMPLETE` while on the successful odd attempts it's equal to `ItemStatus.IDLE`

**Comment by rikschennink** [comment]
@rovansteen I misread your earlier report as every odd file didn't process, rereading it now I notice it's the odd edit that's the problem. I'm on it.

**Comment by rovansteen** [comment]
@rikschennink I don't see `3.7.4` yet. https://www.npmjs.com/package/filepond

**Comment by rikschennink** [comment]
@rovansteen I forgot to publish to npm... should be there any minute now.

**Comment by rovansteen** [comment]
@rikschennink working all fine now, thanks for your help and persistence with this issue! 😄

**Comment by farhadnowzari** [comment]
> Indeed FilePond does not throw an event in that situation as the file is not being processed till it's sent to the server.
> 
> I've updated both FilePond core (to `3.7.2`) and the FilePond Image Edit plugin (to `1.0.2`).
> 
> You can now listen to the Doka `onconfirm` and `oncancel` callbacks. These callbacks receive an additional parameter (over the default Doka callbacks) that contains a reference to the item being edited.
> 
> ```js
> FilePond.create(
>     document.querySelector('input'), 
>     {
>         imageEditEditor: Doka.create({
>             onconfirm: (output, item) => {
>                 console.log('Confirm callback', output, item);
>             },
>             oncancel: (item) => {
>                 console.log('Cancel callback', item);
>             }
>         })
>     }
> );
> ```
Hi,
I know this issue is closed.
But here when doka is calling onconfirm, the item is not the edited item. But I've set allowProcess on filePond to false maybe that is why. but I want to get the cropped image and then add it as a formdata somewhere else to upload.
so the onconfirm returns `output` and `item`. `item` for me is the unedited image and `output` is the crop information without any file. it seems at the end if I had the process on it could set the file property of the output to the edited file but I don't want to process it (I don't want to send it anywhere yet).

**Comment by rikschennink** [comment]
@farhadnowzari Why then use FilePond? If you use FilePond it will handle the processing of the image, it will automatically set `outputFile` to `false` on the Doka instance.

**Comment by farhadnowzari** [comment]
> @farhadnowzari Why then use FilePond? If you use FilePond it will handle the processing of the image, it will automatically set `outputFile` to `false` on the Doka instance.

It is fine, the reason I'm doing that is that I have a vue component which can work standalone and send data to the given endpoint and if the endpoint is not provided it syncs the cropped image back.

I solved it. I found that instead of calling processFile I can call prepareFile and then I can get the onpreparefile(@preparefile) will be called and it has the output with cropped image. :)
